src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,153 @@
+                    return;
+                }
+            }
+        }
+
+        // Cast to a uniform int structure
+        ret.cast_to(Uniform {
+            unit: bits_to_int_reg(bits),
+            total: size
+        });
     } else {
         ret.make_indirect();
-        *offset += cx.tcx.data_layout.pointer_size;
     }
 }
 
-fn classify_arg_ty(cx: &CodegenCx, arg: &mut ArgType, offset: &mut Size) {
+fn classify_arg_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, arg: &mut ArgType<'tcx>) {
     let dl = &cx.tcx.data_layout;
     let size = arg.layout.size;
-    let align = arg.layout.align.max(dl.i32_align).min(dl.i64_align);
 
     if arg.layout.is_aggregate() {
-        arg.cast_to(Uniform {
-            unit: Reg::i64(),
+        let mut chunks = Vec::new();
+        match arg.layout.fields {
+            layout::FieldPlacement::Array { .. } => {
+                // Arrays are passed indirectly
+                arg.make_indirect();
+                return;
+            }
+            layout::FieldPlacement::Union(_) => {
+                // Unions and are always treated as a series of 64-bit integer chunks
+            },
+            layout::FieldPlacement::Arbitrary { .. } => {
+                // Structures are split up into a series of 64-bit integer chunks, but any aligned
+                // doubles not part of another aggregate are passed as floats.
+                let mut last_offset = Size::from_bytes(0);
+
+                for i in 0..arg.layout.fields.count() {
+                    let field = arg.layout.field(cx, i);
+                    let offset = arg.layout.fields.offset(i);
+
+                    // We only care about aligned doubles
+                    if let layout::Abi::Scalar(ref scalar) = field.abi {
+                        if let layout::F64 = scalar.value {
+                            if offset.is_abi_aligned(dl.f64_align) {
+                                // Insert enough integers to cover [last_offset, offset)
+                                assert!(last_offset.is_abi_aligned(dl.f64_align));
+                                for _ in 0..((offset - last_offset).bits() / 64) {
+                                    chunks.push(RegKind::Integer);
+                                }
+
+                                chunks.push(RegKind::Float);
+                                last_offset = offset + Reg::f64().size;
+                            }
+                        }
+                    }
+                }
+            }
+        };
+
+        // Extract first 8 chunks as the prefix
+        let mut prefix = [RegKind::Integer; 8];
+        for (i, &chunk) in chunks.iter().take(8).enumerate() {
+            prefix[i] = chunk
+        }
+
+        arg.cast_to(CastTarget::ChunkedPrefix {
+            prefix: prefix,
+            chunk: Size::from_bytes(8),
             total: size
         });
-        if !offset.is_abi_aligned(align) {
```

Comment:
There's no alignment padding anymore? I'm not even sure what clang does wrt that for other targets - especially since there's a `stackalign` attribute now.

Reply:
I think this is correct. In 32-bit MIPS, we have to pad 64-bit aligned structures manually because we cast to an i32 Uniform so LLVM doesn't know that the structure is misaligned. In 64-bit MIPS we always use i64 chunks so this issue doesn't arise.